src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,145 @@
+            tail = head;
+            size++;
+        } else {
+            Node<T> newNode = new Node<>(value, tail, null);
+            tail.next = newNode;
+            tail = newNode;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node selectedNode = getNodeById(index);
+        if (selectedNode == head) {
+            Node newNode = new Node<>(value, null, selectedNode);
+            selectedNode.prev = newNode;
+            head = newNode;
+        } else {
+            Node newNode = new Node<>(value, selectedNode.prev, selectedNode);
+            selectedNode.prev.next = newNode;
+            selectedNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return (T) getNodeById(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node currentNode = getNodeById(index);
+        T oldValue = (T) currentNode.item;
+        currentNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node nodeToDelete = getNodeById(index);
+        unlink(nodeToDelete);
+        size--;
+        return (T) nodeToDelete.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if ((currentNode.item == object)
+                    || (currentNode.item != null && currentNode.item.equals(object))) {
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlink(Node nodeToDelete) {
+        if (head == nodeToDelete) {
+            if (head != tail) {
+                head = nodeToDelete.next;
+                nodeToDelete.next.prev = null;
+            }
+            nodeToDelete.next = null;
+        } else if (tail == nodeToDelete) {
+            tail = nodeToDelete.prev;
+            nodeToDelete.prev.next = null;
+            nodeToDelete.prev = null;
+        } else {
+            nodeToDelete.prev.next = nodeToDelete.next;
+            nodeToDelete.next.prev = nodeToDelete.prev;
+            nodeToDelete.prev = null;
+            nodeToDelete.next = null;
+        }
+    }
+
+    private Node getNodeById(int index) {
+        if (size / 2 > index) {
+            Node<T> currentNode = tail;
```

Comment:
If your index smaller then size/2, you have to start from head, case its mean that item you search is placed in first half